File: src/Collections/array.ts

import { Wrapped, Thunk, From } from '../../types.ts'


export function New<V>(val: Wrapped<V>, size: Wrapped<number>): Thunk<V[]> {
  return () => {
    const data: V[] = []

    const tgt = From(size)
    for (let idx = 0; idx < tgt; idx++) {
      data[idx] = From(val)
    }

    return data
  }
}


export function Repeat<T>(val: Wrapped<T>, size: Wrapped<number>): Thunk<T[]> {
  return (): T[] => {
    let parts: T[] = []

    let tgt = From(size)
    for (let idx = 0; idx < tgt; idx++) {
      parts[idx] = From(val)
    }

    return parts
  }
}

File: collection/array/array.go
package array

import (
	. "github.com/rgrannell1/peach/types"
)

/*
 * Construct an array of values
 */
func New[T any](val Wrapped[T], size Wrapped[int]) Wrapped[[]T] {
	return Thunk(func () []T {
		tgt := size.From()
		elems := make([]T, tgt)

		for idx := 0; idx < tgt; idx++ {
			elems[idx] = val.From()
		}

		return elems
	})
}
